File: python/Spiral Matrix II.py
class Solution(object):
    def generateMatrix(self, n):
        if n == 0:
            return []
        if n == 1:
            return [[1]]

        matrix = [[-1 for x in range(n)] for y in range(n)]
        direction = right
        current = 2
        pos_x = 0
        pos_y = 0
        matrix[0][0] = 1
        while current <= n * n:
            if direction == right:
                if pos_y + 1 >= n or matrix[pos_x][pos_y + 1] != -1:
                    direction = down
                else:
                    pos_y += 1
                    matrix[pos_x][pos_y] = current
                    current += 1
            elif direction == down:
                if pos_x + 1 >= n or matrix[pos_x + 1][pos_y] != -1:
                    direction = left
                else:
                    pos_x += 1
                    matrix[pos_x][pos_y] = current
                    current += 1
            elif direction == left:
                if pos_y - 1 < 0 or matrix[pos_x][pos_y - 1] != -1:
                    direction = up
                else:
                    pos_y -= 1
                    matrix[pos_x][pos_y] = current
                    current += 1
            elif direction == up:
                if pos_x - 1 < 0 or matrix[pos_x - 1][pos_y] != -1:
                    direction = right
                else:
                    pos_x -= 1
                    matrix[pos_x][pos_y] = current
                    current += 1
        
        return matrix

File: cpp/Spiral Matrix II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> ret1;
        if (n < 1)
            return ret1;

        vector<vector<int>> ret(n, vector<int>(n, 0));
        int total = n * n;
        int i = 0, j = 0, element = 1;
        int rightBound = n - 1;
        int downBound  = n - 1;
        int leftBound  = 0;
        int upBound    = 0;

        while(element <= total)
        {
            //toward right
            if (upBound != 0)
            {
                i++;
                j++;
            }
            for (; j <= rightBound && element <= total; j++)
                ret[i][j] = element++;
            upBound++;//update upBound

            //toward down
            j--;
            i++;
            for (; i <= downBound && element <= total; i++)
                ret[i][j] = element++;
            rightBound--;//update rightBound

            //toward left
            i--;
            j--;
            for (; j >= leftBound && element <= total; j--)
                ret[i][j] = element++;
            downBound--;//update downBound

            //toward up
            j++;
            i--;
            for (; i >= upBound && element <= total; i--)
                ret[i][j] = element++;
            leftBound++;//update leftBound
        }

        return ret;
    }
};
